main/adapter-commons/src/service.ts

```diff
@@ -1,84 +1,206 @@
+ * `__find`, `__get`, `__update`, `__patch` and `__remove` methods.
+ */
+export abstract class AdapterBase<
+  T = any,
+  D = Partial<T>,
+  P extends AdapterParams = AdapterParams,
+  O extends AdapterServiceOptions = AdapterServiceOptions
+> implements InternalServiceMethods<T, D, P>
+{
+  options: O;
+
+  constructor(options: O) {
+    this.options = {
+      id: "id",
+      events: [],
+      paginate: false,
+      multi: false,
+      filters: {},
+      operators: [],
+      ...options,
+    };
   }
-  whitelist?: string[];
-  allow: string[];
-  filters: string[];
-}
 
-export interface AdapterOptions<M = any> extends Pick<ServiceOptions, 'multi'|'allow'|'paginate'> {
-  Model?: M;
-}
+  get id() {
+    return this.options.id;
+  }
 
-export interface AdapterParams<M = any> extends Params {
-  adapter?: Partial<AdapterOptions<M>>;
-}
+  get events() {
+    return this.options.events;
+  }
 
-/**
- * Hook-less (internal) service methods. Directly call database adapter service methods
- * without running any service-level hooks. This can be useful if you need the raw data
- * from the service and don't want to trigger any of its hooks.
- *
- * Important: These methods are only available internally on the server, not on the client
- * side and only for the Feathers database adapters.
- *
- * These methods do not trigger events.
- *
- * @see {@link https://docs.feathersjs.com/guides/migrating.html#hook-less-service-methods}
- */
-export interface InternalServiceMethods<T = any, D = Partial<T>> {
+  /**
+   * Check if this adapter allows multiple updates for a method.
+   * @param method The method name to check.
+   * @param params The service call params.
+   * @returns Wether or not multiple updates are allowed.
+   */
+  allowsMulti(method: string, params: P = {} as P) {
+    const always = alwaysMulti[method];
+
+    if (typeof always !== "undefined") {
+      return always;
+    }
+
+    const { multi } = this.getOptions(params);
+
+    if (multi === true || multi === false) {
+      return multi;
+    }
+
+    return multi?.includes(method);
+  }
 
   /**
-   * Retrieve all resources from this service, skipping any service-level hooks.
+   * Returns the combined options for a service call. Options will be merged
+   * with `this.options` and `params.adapter` for dynamic overrides.
+   *
+   * @param params The parameters for the service method call
+   * @returns The actual options for this call
+   */
+  getOptions(params: P): O {
+    const paginate =
+      params.paginate !== undefined ? params.paginate : this.options.paginate;
+
+    return {
+      ...this.options,
+      paginate,
+      ...params.adapter,
+    };
+  }
+
+  /**
+   * Sanitize the incoming data, e.g. removing invalid keywords etc.
+   *
+   * @param data The data to sanitize
+   * @param _params Service call parameters
+   * @returns The sanitized data
+   */
+  async sanitizeData<X = Partial<D>>(data: X, _params?: P) {
```

Comment:
For typescript I had to make `_params` optional as in the `_create` method, `sanitizeData` is called with params that are maybe undefined